src/planner.cc

```diff
@@ -310,201 +381,145 @@ double Planner::Impl::euclidian_speed(const struct AxisTarget *t) {
 
-  if (do_accel && accel_fraction * abs_defining_axis_steps > 0) {
-    has_accel = true;
-    accel_command.v0 = (float)last_speed;         // Last speed of defining axis
-    accel_command.v1 = (float)target_pos->speed;  // New speed of defining axis
-
-    // Now map axis steps to actual motor driver
-    for (const GCodeParserAxis a : AllAxes()) {
-      const int accel_steps = std::lround(accel_fraction * axis_steps[a]);
-      assign_steps_to_motors(&accel_command, a, accel_steps);
+  // 2) submit the segments
+  // Flush the full queue.
+  struct LinearSegmentSteps accel_command = {};
+  struct LinearSegmentSteps move_command = {};
+  struct LinearSegmentSteps decel_command = {};
+  PlanningSegment *segment = NULL;
+
+  for (uint32_t i = 0; i < num_segments; ++i) {
+    segment = planning_buffer_[0];
+
+    memset(&move_command, 0, sizeof(move_command));
+    move_command.aux_bits = segment->target.aux_bits;
+    memcpy(&accel_command, &move_command, sizeof(accel_command));
+    memcpy(&decel_command, &move_command, sizeof(decel_command));
+
+    const unsigned defining_axis_steps = segment->planned.TotalSteps();
+    const double accel_fraction =
+      (double)segment->planned.accel / defining_axis_steps;
+    const double decel_fraction =
+      (double)segment->planned.decel / defining_axis_steps;
+
+    // Accel
+    if (segment->planned.accel) {
+      for (const GCodeParserAxis a : AllAxes()) {
+        const int accel_steps =
+          std::lround(accel_fraction * segment->target.delta_steps[a]);
+        assign_steps_to_motors(&accel_command, a, accel_steps);
+      }
+      accel_command.v0 = segment->planned.v0;
+      accel_command.v1 = segment->planned.v1;
     }
-  } else {
-    if (last_speed)
-      target_pos->speed = last_speed;  // No accel so use the last speed
-  }
-
-  move_command.v0 = (float)target_pos->speed;
-  move_command.v1 = (float)target_pos->speed;
 
-  if (do_accel && decel_fraction * abs_defining_axis_steps > 0) {
-    has_decel = true;
-    decel_command.v0 = (float)target_pos->speed;
-    decel_command.v1 = (float)next_speed;
-    target_pos->speed = next_speed;
+    // Decel
+    if (segment->planned.decel) {
+      for (const GCodeParserAxis a : AllAxes()) {
+        const int decel_steps =
+          std::lround(decel_fraction * segment->target.delta_steps[a]);
+        assign_steps_to_motors(&decel_command, a, decel_steps);
+      }
+      decel_command.v0 = segment->planned.v1;
+      decel_command.v1 = segment->planned.v2;
+    }
 
-    // Now map axis steps to actual motor driver
+    // Travel
     for (const GCodeParserAxis a : AllAxes()) {
-      const int decel_steps = std::lround(decel_fraction * axis_steps[a]);
-      assign_steps_to_motors(&decel_command, a, decel_steps);
+      assign_steps_to_motors(&move_command, a,
+                             segment->target.delta_steps[a] -
+                               accel_command.steps[a] - decel_command.steps[a]);
```

Comment:
`accel_command` and `decel_command` are of type `LinearSegmentSteps`.
Their `steps` array only holds `BEAGLEG_NUM_MOTORS` elements, not the number covered by all the axes.
So here is a mix-up between axis and motor.

Reply:
nice catch! (solved)